### PR TITLE
sst_importer: add write interface to remove importer for lightning

### PR DESCRIPTION
**Pull request by 3pointer**
### What problem does this PR solve?

Issue Number: close https://github.com/tikv/tikv/issues/7404
Problem Summary:
Add write interface to generate sst from streaming with sorted kv.

### What is changed and how it works?
1.first update kvproto https://github.com/pingcap/kvproto/pull/602 add new write streaming interface
2. parse SstMeta just like download. build Default and Write SSTWriter with this meta.
3. transform KV to default CF and Write CF to generate sst file.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
1. build a go client in lightning.
2. try send sorted kv to TiKV.
3. generate two sst files with these kvs( xxx_default.sst, xxx_write.sst)

Side effects

### Release note

### CODE REVIEW

File: src/import/sst_service.rs [hunk 1]
```diff
@@ -360,4 +360,64 @@ impl<Router: RaftStoreRouter> ImportSst for ImportSSTService<Router> {
...
+                                .set_db(&engine)
+                                .set_cf(CF_DEFAULT)
+                                .build(&name.to_str().unwrap())?;
+                            let write = <RocksEngine as SstExt>::SstWriterBuilder::new()
+                                .set_in_memory(true)
+                                .set_db(&engine)
+                                .set_cf(CF_WRITE)
+                                .build(&name.to_str().unwrap())?;
+                            let mut writer =
+                                import.new_writer::<RocksEngine>(default, write, &meta)?;
```
Comment (kennytm) [inline src/import/sst_service.rs:410]: The SST writer should be created *outside* of the `for_each` loop.

File: components/sst_importer/src/sst_importer.rs [hunk 2]
```diff
@@ -348,6 +349,129 @@ impl SSTImporter {
...
+            write,
+            default_path,
+            write_path,
+            default_meta,
+            write_meta,
+        ))
+    }
+}
+
+pub struct SSTWriter<E: KvEngine> {
```
Comment (kennytm) [inline components/sst_importer/src/sst_importer.rs:380]: this probably doesn't need to be `pub`.

File: components/sst_importer/src/sst_importer.rs [hunk 3]
```diff
@@ -348,6 +349,129 @@ impl SSTImporter {
...
+    pub fn finish(self) -> Result<Vec<SstMeta>> {
+        let default_meta = self.default_meta.clone();
+        let write_meta = self.write_meta.clone();
+        let mut metas = vec![];
+        let (default_entries, write_entries) = (self.default_entries, self.write_entries);
+        let (p1, p2) = (self.default_path.clone(), self.write_path.clone());
+        let (w1, w2) = (self.default, self.write);
+        if default_entries > 0 {
+            let (_, sst_reader) = w1.finish_read()?;
+            SSTWriter::<E>::save(sst_reader, p1)?;
```
Comment (kennytm) [inline components/sst_importer/src/sst_importer.rs:447]:
```suggestion
            Self::save(sst_reader, p1)?;
```

File: components/sst_importer/src/sst_importer.rs [hunk 4]
```diff
@@ -348,6 +349,129 @@ impl SSTImporter {
...
+            metas.push(default_meta);
+        }
+        if write_entries > 0 {
+            let (_, sst_reader) = w2.finish_read()?;
+            SSTWriter::<E>::save(sst_reader, p2)?;
```
Comment (kennytm) [inline components/sst_importer/src/sst_importer.rs:452]:
```suggestion
            Self::save(sst_reader, p2)?;
```

File: components/sst_importer/src/sst_importer.rs [hunk 5]
```diff
@@ -348,6 +349,129 @@ impl SSTImporter {
...
+            metas.push(write_meta);
+        }
+        info!(
+            "finish write to sst with default entries: {}, write entries: {}",
+            default_entries, write_entries
+        );
```
Comment (kennytm) [inline components/sst_importer/src/sst_importer.rs:458]: please use structured logging.

File: components/sst_importer/src/sst_importer.rs [hunk 6]
```diff
@@ -348,6 +349,129 @@ impl SSTImporter {
...
+            self.default.put(&k, value)?;
+            self.default_entries += 1;
+        }
+        Ok(())
+    }
+
+    pub fn finish(self) -> Result<Vec<SstMeta>> {
+        let default_meta = self.default_meta.clone();
+        let write_meta = self.write_meta.clone();
+        let mut metas = vec![];
```
Comment (kennytm) [inline components/sst_importer/src/sst_importer.rs:441]:
```suggestion
        let mut metas = Vec::with_capacity(2);
```

File: components/sst_importer/src/sst_importer.rs [hunk 7]
```diff
@@ -348,6 +349,129 @@ impl SSTImporter {
...
+        Ok(metas)
+    }
+
+    fn save(
+        sst_reader: <<E as SstExt>::SstWriter as SstWriter>::ExternalSstFileReader,
+        import_path: ImportPath,
+    ) -> Result<()> {
+        let tmp_path = import_path.temp;
+        let mut tmp_f = AllowStdIo::new(File::create(&tmp_path)?);
+        block_on(copy(AllowStdIo::new(sst_reader), &mut tmp_f))?;
```
Comment (kennytm) [inline components/sst_importer/src/sst_importer.rs:468]: why use async?

File: src/import/sst_service.rs [hunk 8]
```diff
@@ -360,4 +366,79 @@ impl<Router: RaftStoreRouter<RocksEngine>> ImportSst for ImportSSTService<Router
...
+                                return Err(Error::InvalidChunk);
+                            }
+                        };
+                        Ok((writer, stream))
+                    })
+                    .and_then(move |(writer, stream)| {
+                        stream
+                            .map_err(Error::from)
+                            .fold(writer, |mut writer, req| {
+                                let _start = Instant::now_coarse();
```
Comment (kennytm) [inline src/import/sst_service.rs:422]: ?

File: components/sst_importer/src/sst_importer.rs [hunk 9]
```diff
@@ -348,6 +349,129 @@ impl SSTImporter {
...
+    pub fn new_writer<E: KvEngine>(
+        &self,
+        default: E::SstWriter,
+        write: E::SstWriter,
+        meta: SstMeta,
+    ) -> Result<SSTWriter<E>> {
+        // use new uuid to generate different default and write filename
+        let mut default_meta = meta.clone();
+        let cf = CF_DEFAULT.to_owned();
+        default_meta.set_cf_name(cf);
```
Comment (kennytm) [inline components/sst_importer/src/sst_importer.rs:362]:
```suggestion
        default_meta.set_cf_name(CF_DEFAULT.to_owned());
```

File: components/sst_importer/src/sst_importer.rs [hunk 10]
```diff
@@ -348,6 +349,129 @@ impl SSTImporter {
...
+        let default_path = self.dir.join(&default_meta)?;
+
+        let mut write_meta = meta;
+        let cf = CF_WRITE.to_owned();
+        write_meta.set_cf_name(cf);
```
Comment (kennytm) [inline components/sst_importer/src/sst_importer.rs:367]:
```suggestion
        write_meta.set_cf_name(CF_WRITE.to_owned());
```

File: components/sst_importer/src/sst_importer.rs (hunk 2 above)
Comment (3pointer) [reply components/sst_importer/src/sst_importer.rs:376]: we will build SSTWriter in src/import/sst_services.rs, so this need to be public.

Comment (kennytm) [reply components/sst_importer/src/sst_importer.rs:376]: Okay

File: components/sst_importer/src/sst_importer.rs [hunk 11]
```diff
@@ -348,6 +348,126 @@ impl SSTImporter {
...
+        Ok(metas)
+    }
+
+    fn save(
+        sst_reader: <<E as SstExt>::SstWriter as SstWriter>::ExternalSstFileReader,
+        import_path: ImportPath,
+    ) -> Result<()> {
+        let tmp_path = import_path.temp;
+        let mut tmp_f = AllowStdIo::new(File::create(&tmp_path)?);
+        std::io::copy(&mut AllowStdIo::new(sst_reader), &mut tmp_f)?;
```
Comment (kennytm) [inline components/sst_importer/src/sst_importer.rs:464]:
```suggestion
        let mut tmp_f = File::create(&tmp_path)?;
        std::io::copy(&mut sst_reader, &mut tmp_f)?;
```

File: tests/integrations/import/sst_service.rs [hunk 12]
```diff
@@ -78,6 +78,27 @@ fn test_upload_sst() {
...
+    meta.set_region_id(ctx.get_region_id());
+    meta.set_region_epoch(ctx.get_region_epoch().clone());
+
+    let mut keys = vec![];
+    let mut values = vec![];
+    for i in 1..10 {
+        keys.push(format!("k{}", i.to_string()).into_bytes());
+        values.push(format!("v{}", i.to_string()).into_bytes());
+    }
+    send_write_sst(&import, &meta, keys, values).unwrap();
```
Comment (kennytm) [inline tests/integrations/import/sst_service.rs:99]: nothing is checked in this test. i expect you to

1. write the SST files
2. ingest the two SST files
3. perform `check_ingested_kvs`

Use `test_ingest_sst_without_crc32` as sample.

**Comment by kennytm** [comment]
/merge

yet another copr-test failure https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tikv_ghpr_integration-copr-test/detail/tikv_ghpr_integration-copr-test/2610/pipeline

```
[2020-04-28T07:39:12.506Z] Statement: #819 -  SELECT STRCMP( '23:10:42.007765', `col_bigint_unsigned` ) AS field1, `col_varbinary_32` NOT LIKE `col_set` AS field2, `col_tinyint_unsigned_key` IS FALSE AS field3, ( INTERVAL( `col_bit`, ( INTERVAL( '2018-04-26 10:38:03.031949', `col_char_2`, NULL ) ), `col_varchar_1_key` ) ) NOT BETWEEN NULL AND NULL AS field4, COALESCE( '2018-04-27', -26901, ( `col_float_unsigned_key` <=> ( `col_float_unsigned_key` > 'pcbqbfmwpebbkyfsxemhlybhtahsdfttztbnrqjpdtwjharagadcroqtxlefjrhcokdymxvanwfvayfdbhtoxwppiavhgmizbrrxgafhbcxkosudjiyckygmuatynejcqwwbclkmfhgrrenyxlyawqekkchtjzebuphvkwswxbtqsjokzalmxfaklbeukgslyqnrheytuhbqsbseiojnyxesmnsdfyyisxjoljtdmdxwycmyxfnxnojmst' ) ), `col_binary_8_key`, `col_char_255` ) AS field5 FROM `table1000_int_autoinc` WHERE COALESCE( `col_blob`, '2030-01-04', `col_decimal_key` ) ORDER BY field1, field2, field3, field4, field5 LIMIT 3 /* QNO 824 CON_ID 112 */ ;
```

**Pull request merged**